crates/service/src/algorithms/clustering/elkan_k_means.rs

```diff
@@ -24,30 +30,36 @@ impl<S: G> ElkanKMeans<S> {
 
         let mut rand = StdRng::from_entropy();
         let mut centroids = Vec2::new(dims, c);
-        let mut lowerbound = Square::new(n, c);
+        let lowerbound = SyncUnsafeCell::new(Square::new(n, c));
         let mut upperbound = vec![F32::zero(); n];
         let mut assign = vec![0usize; n];
 
         centroids[0].copy_from_slice(&samples[rand.gen_range(0..n)]);
 
-        let mut weight = vec![F32::infinity(); n];
+        let weight = SyncUnsafeCell::new(vec![F32::infinity(); n]);
         for i in 0..c {
             let mut sum = F32::zero();
-            for j in 0..n {
+            (0..n).into_par_iter().for_each(|j| {
                 let dis = S::elkan_k_means_distance(&samples[j], &centroids[i]);
-                lowerbound[(j, i)] = dis;
-                if dis * dis < weight[j] {
-                    weight[j] = dis * dis;
+                unsafe {
+                    (&mut *lowerbound.get())[(j, i)] = dis;
```

Comment:
It's absolutely unsound. Please use a structure that is like `Square<Atomic<F32>>`.